[Feature] Add logprobs support for Whisper transcription API

TheCodeWrangler opened the pull request:
## Purpose

Add `logprobs` parameter support to the `/v1/audio/transcriptions` endpoint, enabling users to retrieve log probabilities for transcribed tokens.

This is a vLLM extension to the OpenAI Whisper API (OpenAI's Whisper API does not support logprobs).

**Features:**
- New `logprobs` parameter in `TranscriptionRequest` - specify number of top alternatives to return
- Response includes `TranscriptionLogProbs` with:
  - `tokens`: List of generated tokens
  - `token_logprobs`: Log probability of each token  
  - `top_logprobs`: Dictionary of top N alternative tokens and their log probabilities
- Works with both streaming and non-streaming responses
- Works with both `json` and `verbose_json` response formats

## Test Plan

```bash
# Start server with Whisper model
vllm serve openai/whisper-large-v3-turbo

# Test with logprobs
python -c "
from openai import OpenAI
client = OpenAI(base_url='http://localhost:8000/v1', api_key='test')

with open('audio.mp3', 'rb') as f:
    result = client.audio.transcriptions.create(
        model='openai/whisper-large-v3-turbo',
        file=f,
        extra_body={'logprobs': 5}
    )
print(result.logprobs)
"
```

Run unit tests:
```bash
pytest tests/entrypoints/openai/test_transcription_validation.py -v -k logprobs
```

## Test Result

```
=== Test with logprobs ===
Time taken: 0.17 seconds
Text:  Okay, I'm just going to record a 30 second chunk here...
                                   right 0.9994
                           <|endoftext|> 0.0005
--
                                       , 0.7603
                                       . 0.2392
--
                                      we 0.9982
                                      25 0.0004
...
```

Logprobs are correctly returned with top-k alternatives for each token position.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
  - Updated `docs/serving/openai_compatible_server.md` with vLLM Extensions section documenting logprobs
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

---

> [!NOTE]
> Introduces vLLM `logprobs` support for Whisper-style transcriptions with OpenAI-compatible response shapes.
> 
> - Protocol: adds `logprobs` to `TranscriptionRequest` sampling params; defines `TranscriptionLogprob`; includes `logprobs` in `TranscriptionResponse`, `TranscriptionResponseVerbose`, and streaming `TranscriptionResponseStreamChoice`.
> - Server: implements logprob extraction/formatting in `speech_to_text.py` (non‑streaming and streaming, `json` and `verbose_json`), with tokenizer-based decoding and validation.
> - Docs: updates `docs/serving/openai_compatible_server.md` with vLLM Extensions section and example usage of `extra_body={"logprobs": N}`.
> - Tests: adds validation tests covering with/without `logprobs` and `verbose_json` structure.
> - Runtime: warms up sampler both without and with logprobs in `gpu_model_runner.py` to precompile logprobs path.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4704208ae6a9ad747c9ecc40fb230e2b7d9438a3. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> <sup>[Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) is generating a summary for commit 8d7609c2e581e39ab47fc04649e58db301197d86. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

TheCodeWrangler committed:
WIP initial test

Signed-off-by: Nathan Price <nathan@abridge.com>

TheCodeWrangler committed:
Updated docs

Signed-off-by: Nathan Price <nathan@abridge.com>

TheCodeWrangler committed:
Add logprobs tests and fix forward reference

Signed-off-by: Nathan Price <nathan@abridge.com>

TheCodeWrangler committed:
Fix mypy type error in logprobs

Signed-off-by: Nathan Price <nathan@abridge.com>

TheCodeWrangler committed:
Clarify logprobs returns N+1 tokens (consistent with completion API)

Signed-off-by: Nathan Price <nathan@abridge.com>

TheCodeWrangler committed:
Fix mypy errors: use getattr for logprobs on union type

Signed-off-by: Nathan Price <nathan@abridge.com>

DarkLight1337 commented on vllm/entrypoints/openai/speech_to_text.py:
```diff
@@ -369,6 +372,61 @@ def _get_verbose_segments(
+        for i, token_id in enumerate(token_ids):
+            step_top_logprobs = top_logprobs[i]
+            if step_top_logprobs is None:
+                token = self.tokenizer.decode(token_id)
```
`self.tokenizer` may be `None`, I suggest following the logic of `_create_completion_logprobs` and pass `tokenizer` as an argument so that you can validate it here.

TheCodeWrangler committed:
refactor: Align transcription logprobs with OpenAI format

Change TranscriptionLogProbs to match OpenAI's Logprob type from
openai.types.audio.transcription. The response now returns a list of
TranscriptionLogprob objects, each containing:
- token: The token string
- bytes: UTF-8 byte encoding of the token
- logprob: The log probability

This replaces the previous parallel-arrays format (tokens, token_logprobs,
top_logprobs) with the per-token object format used by OpenAI.

Signed-off-by: Nathan Price <nathan@abridge.com>

TheCodeWrangler committed:
fix: Validate tokenizer in _create_transcription_logprobs

Pass tokenizer as an argument and validate it is not None before use,
following the pattern from _create_completion_logprobs. Raises
VLLMValidationError if tokenizer is None (skip_tokenizer_init=True).

Signed-off-by: Nathan Price <nathan@abridge.com>

TheCodeWrangler committed:
fix(warmup): warm up sampler logprobs path to avoid first-call JIT

The sampler warmup previously only exercised the non-logprobs path.
When a request with logprobs=N was made, the first call triggered
torch.compile JIT compilation for batched_count_greater_than, causing
significant latency (~12s) on the first logprobs request.

This change runs the sampler warmup twice: once without logprobs and
once with logprobs enabled, ensuring the torch.compile JIT compilation
happens during engine initialization rather than on the first request.

Signed-off-by: Nathan Price <nathan@abridge.com>

TheCodeWrangler committed:
Merge branch 'main' into nathan/whisper-logprobs

TheCodeWrangler committed:
Merge branch 'main' into nathan/whisper-logprobs

TheCodeWrangler committed:
Merge branch 'main' into nathan/whisper-logprobs

TheCodeWrangler committed:
Merge branch 'main' into nathan/whisper-logprobs

TheCodeWrangler committed:
Merge branch 'main' into nathan/whisper-logprobs

TheCodeWrangler committed:
Merge branch 'main' into nathan/whisper-logprobs

TheCodeWrangler committed:
Merge branch 'main' into nathan/whisper-logprobs

TheCodeWrangler committed:
Merge branch 'main' into nathan/whisper-logprobs

NickLucche commented:
Related to https://github.com/vllm-project/vllm/pull/31059, which still follows OAI specs though. 
I am afraid that adding logprobs to the default response format too would bring in extra replication as user can just switch to verbose_json as per oai docs @TheCodeWrangler @DarkLight1337
